### PR TITLE
Feature/gray matter

### PR DESCRIPTION
**Pull request by Eransho**
Introducing inline completions feature

### CODE REVIEW

File: src/main/java/com/tabnine/inline/TabnineDocumentListener.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+public class TabnineDocumentListener implements DocumentListener {
+
+  private CompletionPreview preview;
+
+  private static final AtomicBoolean isMuted = new AtomicBoolean(false);
+
+  @Override
+  public void documentChanged(@NotNull DocumentEvent event) {
+    if (isMuted.get()
```
Comment (Eransho) [inline src/main/java/com/tabnine/inline/TabnineDocumentListener.java:32]: @dimacodota flow starts here

File: src/main/java/com/tabnine/inline/InlineCompletionHandler.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+
+import static com.tabnine.general.StaticConfig.MAX_COMPLETIONS;
+import static com.tabnine.general.Utils.endsWithADot;
+
+public class InlineCompletionHandler implements CodeInsightActionHandler {
+  private static final String INLINE_DUMMY_IDENTIFIER = "TabnineInlineDummy";
```
Comment (Eransho) [inline src/main/java/com/tabnine/inline/InlineCompletionHandler.java:37]: @dimacodota main class for handling the inline from all 3 use cases: typing, next and prev actions

File: src/main/java/com/tabnine/inline/CompletionPreview.java [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+import java.awt.event.KeyAdapter;
+import java.awt.event.KeyEvent;
+import java.awt.event.KeyListener;
+import java.awt.font.TextAttribute;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class CompletionPreview implements Disposable {
```
Comment (Eransho) [inline src/main/java/com/tabnine/inline/CompletionPreview.java:40]: This class is responsible for the UI part (i.e. the gray matter)

File: src/main/java/com/tabnine/inline/TabnineDocumentListener.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    PsiFile file =
+        ObjectUtils.doIfNotNull(
+            project, proj -> PsiDocumentManager.getInstance(proj).getPsiFile(document));
+    if (editor == null || project == null || file == null) {
+      return;
+    }
+    if (preview == null) {
+      preview = CompletionPreview.findOrCreateCompletionPreview(editor, file);
+    }
+    InlineCompletionHandler handler = new InlineCompletionHandler(true);
```
Comment (yonip23) [inline src/main/java/com/tabnine/inline/TabnineDocumentListener.java:57]: maybe enum `Forward/Backwards`

Comment (yonip23) [reply src/main/java/com/tabnine/inline/TabnineDocumentListener.java:57]: save handler as data member

File: src/main/java/com/tabnine/inline/InlineCompletionHandler.java [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+      return;
+    }
+
+    // data about previous completions, or just a new fresh state for the editor
+    final CompletionState completionState = CompletionState.findOrCreateCompletionState(editor);
+    int lastDisplayedCompletionIndex = completionState.lastDisplayedCompletionIndex;
+
+    boolean noOldSuggestion = lastDisplayedCompletionIndex == -1 || completionState.prefix == null;
+    boolean editorLocationHasChanged = completionState.lastStartOffset != offset;
+    boolean documentChanged = completionState.lastModCount != document.getModificationStamp();
```
Comment (yonip23) [inline src/main/java/com/tabnine/inline/InlineCompletionHandler.java:65]: `lastModificationStamp`

File: src/main/java/com/tabnine/inline/InlineCompletionHandler.java [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+      @NotNull PsiFile file,
+      CompletionState completionState,
+      int startOffset) {
+    if (completionState.suggestions == null || completionState.suggestions.isEmpty()) {
+      return;
+    }
+    int nextIndex;
+    if (myForward) {
+      nextIndex = completionState.lastDisplayedCompletionIndex + 1;
+      if (nextIndex >= completionState.suggestions.size()) {
```
Comment (yonip23) [inline src/main/java/com/tabnine/inline/InlineCompletionHandler.java:114]: use %

File: src/main/java/com/tabnine/inline/CompletionPreview.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+  public void dispose() {
+    clear();
+    editor.putUserData(INLINE_COMPLETION_PREVIEW, null);
+    CompletionState.clearCompletionState(editor);
+  }
+
+  private void applyPreview() {
+    WriteCommandAction.runWriteCommandAction(
+        file.getProject(),
+        "Tabnine Inline Completion",
```
Comment (yonip23) [inline src/main/java/com/tabnine/inline/CompletionPreview.java:156]: const

File: src/main/java/com/tabnine/inline/CompletionPreview.java [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+                inlay.getOffset() - completions.get(previewIndex).completionPrefix.length();
+            int endOffset = inlay.getOffset() + suffix.length();
+            editor.getDocument().insertString(inlay.getOffset(), suffix);
+            editor.getCaretModel().moveToOffset(endOffset);
+            AutoImporter.registerTabNineAutoImporter(
+                editor, file.getProject(), startOffset, endOffset);
+            previewListener.previewSelected(
+                new CompletionPreviewListener.CompletionPreviewData(
+                    completions, previewIndex, file));
+            Disposer.dispose(CompletionPreview.this);
```
Comment (yonip23) [inline src/main/java/com/tabnine/inline/CompletionPreview.java:171]: catch

File: src/main/java/com/tabnine/capabilities/CapabilitiesService.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public class CapabilitiesService {
+
+  private final BinaryRequestFacade binaryRequestFacade =
+      DependencyContainer.instanceOfBinaryRequestFacade();
+  private final Map<Capability, Boolean> enabledCapabilities = new HashMap<>();
```
Comment (yonip23) [inline src/main/java/com/tabnine/capabilities/CapabilitiesService.java:20]: hash set

File: src/main/java/com/tabnine/inline/TabnineDocumentListener.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+public class TabnineDocumentListener implements DocumentListener {
+
+  private CompletionPreview preview;
+
+  private static final AtomicBoolean isMuted = new AtomicBoolean(false);
+
+  @Override
+  public void documentChanged(@NotNull DocumentEvent event) {
+    if (isMuted.get()
+        || SuggestionsMode.getSuggestionMode() == SuggestionsMode.AUTOCOMPLETE
```
Comment (yonip23) [inline src/main/java/com/tabnine/inline/TabnineDocumentListener.java:33]: != INLINE

File: src/main/java/com/tabnine/lifecycle/LifeCycleHelper.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
+package com.tabnine.lifecycle;
+
+import com.intellij.openapi.Disposable;
+import com.intellij.openapi.components.ServiceManager;
+
+public class LifeCycleHelper implements Disposable {
```
Comment (yonip23) [inline src/main/java/com/tabnine/lifecycle/LifeCycleHelper.java:6]: explain in comment plz

File: src/test/java/com/tabnine/plugin/InlineCompletionTests.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  @Test
+  public void noInlineCompletionWhenAutocompleteSuggestionMode() throws Exception {
+    configureInlineTest(SuggestionsMode.AUTOCOMPLETE);
+
+    type("\nt");
+    assertNull(CompletionPreview.getPreviewText(myFixture.getEditor()));
+  }
+
+  @Test
+  public void showFirstSuggestionWhenExecutingNextInlineAction() throws Exception {
```
Comment (yonip23) [inline src/test/java/com/tabnine/plugin/InlineCompletionTests.java:58]: show second

File: src/test/java/com/tabnine/plugin/InlineCompletionTests.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    type("\nte");
+    myFixture.performEditorAction(ShowNextInlineCompletionAction.ACTION_ID);
+    assertEquals(
+        "Incorrect next inline completion",
+        "mporary",
+        CompletionPreview.getPreviewText(myFixture.getEditor()));
+  }
+
+  @Test
+  public void showLastSuggestionWhenExecutingPrevInlineAction() throws Exception {
```
Comment (yonip23) [inline src/test/java/com/tabnine/plugin/InlineCompletionTests.java:70]: add 1 more test that checks next->next->prev

**Review by yonip23 — APPROVED** [review]
🦈

**Pull request merged**